### PR TITLE
[Core] Improve performance for 'ov::Model::add_output'

### PR DESCRIPTION
**Pull request by nosovmik**
### Details:
On first call of `add_output(tensor_name)` all available tensor names are cached.
Next calls take nodes from cache which significantly reduces complexity.
Cache is invalidated if topological cache is not valid or cache points to incorrect output (no tensor name of this node anymore)

The same caching is done for 'add_output(op_name, output_index)'

Tests:
- Verifies that adding outputs to all nodes has linear complexity O(N), not O(N^2)
- Verifies cache invalidation scenarios

### Tickets:
 - 68616

### CODE REVIEW

File: src/core/include/openvino/core/model.hpp [hunk 1]
```diff
@@ -346,6 +346,9 @@ class OPENVINO_API Model : public std::enable_shared_from_this<Model> {
     // node has no consumers but still exists in a graph.
     mutable std::vector<std::weak_ptr<Node>> m_cached_ordered_ops;
```
Comment (ilyachur) [inline src/core/include/openvino/core/model.hpp:347]: Can we fill unordered maps together with m_cached_ordered_ops?

Do you have any performance measurements?

Comment (nosovmik) [reply src/core/include/openvino/core/model.hpp:347]: I didn't test all possible models from model zoo.
I tested this on synthesized models with 2k-20k nodes. And used 'add_output' to each node. Without the improvement, 1k takes ~1.7 seconds, 10k takes > 200 seconds.
With improvement, all these are completed within less than one second: 10k with O(N) is processed within same time as 100 nodes with O(N*N).

Test that I added verifies performance N1=N and N2=50*N nodes (adding results to each node) and it verifies that `T1*2500` is significantly less than `T2`

File: src/core/src/model.cpp [hunk 2]
```diff
@@ -298,6 +298,8 @@ std::vector<shared_ptr<ov::Node>> ov::Model::get_ordered_ops() const {
         m_cached_ordered_ops.push_back(node);
         node->insert_info(m_shared_rt_info);
     });
+    m_cached_output_names.clear();
```
Comment (ilyachur) [inline src/core/src/model.cpp:301]: Why do you need to clear these vectors in this method?

Comment (nosovmik) [reply src/core/src/model.cpp:301]: Yes, this is needed, but I found that test cases verifying this need are absent. Added appropriate test cases which will be failed if this 'clear' is absent. Brief scenario:
```
// Scenario:
// 1. Create model with nodes A,B,C.
// 2. Add output to some node 'A' - 'names cache' will be created
// 3. Remove some node 'B' - ordered_ops_cache will be 'false'
// 4. Assign name 'B' to existing node 'C'
// 5. Call get_ordered_ops (ordered_ops_cache=true)
// 6. Expect: Add_output to 'B' - output to node 'C' shall be added on step 4
// Without clearing 'names cache' on step 5 - test will incorrectly add output to 'B'
```

**Review by ilyachur — APPROVED** [review]

**Pull request merged**